internal/controller/apiextensions/composite/environment/selector.go

```diff
@@ -106,28 +123,33 @@ func (s *APIEnvironmentSelector) buildEnvironmentConfigRefFromRef(ref *v1.Enviro
 	}
 }
 
-func (s *APIEnvironmentSelector) buildEnvironmentConfigRefFromSelector(ctx context.Context, cr resource.Composite, selector *v1.EnvironmentSourceSelector) (corev1.ObjectReference, error) {
+func (s *APIEnvironmentSelector) buildEnvironmentConfigRefFromSelector(ctx context.Context, cr resource.Composite, selector *v1.EnvironmentSourceSelector) ([]corev1.ObjectReference, error) {
 	matchLabels := make(client.MatchingLabels, len(selector.MatchLabels))
 	for i, m := range selector.MatchLabels {
 		val, err := ResolveLabelValue(m, cr)
 		if err != nil {
-			return corev1.ObjectReference{}, errors.Wrapf(err, errFmtResolveLabelValue, i)
+			return []corev1.ObjectReference{}, errors.Wrapf(err, errFmtResolveLabelValue, i)
 		}
 		matchLabels[m.Key] = val
 	}
 	res := &v1alpha1.EnvironmentConfigList{}
 	if err := s.kube.List(ctx, res, matchLabels); err != nil {
-		return corev1.ObjectReference{}, errors.Wrap(err, errListEnvironmentConfigs)
+		return []corev1.ObjectReference{}, errors.Wrap(err, errListEnvironmentConfigs)
 	}
 	if len(res.Items) == 0 {
-		return corev1.ObjectReference{}, errors.New(errListEnvironmentConfigsNoResult)
+		return []corev1.ObjectReference{}, nil
 	}
-	envConfig := res.Items[0]
-	return corev1.ObjectReference{
-		Name:       envConfig.Name,
-		Kind:       v1alpha1.EnvironmentConfigKind,
-		APIVersion: v1alpha1.SchemeGroupVersion.String(),
-	}, nil
+
+	envConfigs := make([]corev1.ObjectReference, len(res.Items))
+	for i, v := range res.Items {
+		envConfigs[i] = corev1.ObjectReference{
+			Name:       v.Name,
+			Kind:       v1alpha1.EnvironmentConfigKind,
+			APIVersion: v1alpha1.SchemeGroupVersion.String(),
+		}
+	}
```

Comment:
Aren't both the original implementation and this one assuming that List is always going to return EnvironmentConfigs in the same order? Kind of hard to test in practice, but couldn't that potentially lead to flapping, if for example two EnvironmentConfigs are defining the same key with different values? While with the original implementation couldn't it have selected a different one at the next reconciliation loop. Should we enforce some kind of ordering? Not sure which one though.

Moreover, I feel like the [RefForFirstLabelSelectedObject](https://github.com/crossplane/crossplane/blob/master/internal/controller/apiextensions/composite/environment/selector_test.go#L247) test-case should fail with this PR, but it's not 🤔 CC @MisterMX

Reply:
I did some tests and given that the selected environmentConfigsRefs are stored in the XR at `spec.environmentConfigRefs`, it won't flap on a single resource, but it's going to pick an order randomly when a new XR is created. 🤔 Do we want to take the change to enforce an order?

Reply:
`RefForFirstLabelSelectedObject` test-case might have misleading name. since composite has set env config reference:
```go
cr: composite(
	withEnvironmentRefs(environmentConfigRef("test")),
),
```
and resolvePolicy is not set,`isNoOp` fn signals that no further resolution is needed. hence only already present ref is kept. This applies to original version as well which has this check in `SelectEnvironment()`:
```go
// noop if EnvironmentConfig references are already computed
if len(cr.GetEnvironmentConfigReferences()) > 0 {
    return nil
}
```

Reply:
Created #4070 to track that as I feel it's independent to this PR, but could still be addressed here as I feel like we should not select randomly ordered configs at each time, same as we should not just pick one at random as it's currently done.

For the test case, I think it's more likely that the title is correct and the content wrong, the following changes would make it consistent with the test case name, pass with the current implementation and fail with this patch, which means the test case should be updated according to this patch new behavior:
```diff
diff --git a/internal/controller/apiextensions/composite/environment/selector_test.go b/internal/controller/apiextensions/composite/environment/selector_test.go
index 207614ce..e8e94da8 100644
--- a/internal/controller/apiextensions/composite/environment/selector_test.go
+++ b/internal/controller/apiextensions/composite/environment/selector_test.go
@@ -265,9 +265,7 @@ func TestSelect(t *testing.T) {
 						return nil
 					}),
 				},
-				cr: composite(
-					withEnvironmentRefs(environmentConfigRef("test")),
-				),
+				cr: composite(),
 				rev: &v1.CompositionRevision{
 					Spec: v1.CompositionRevisionSpec{
 						Environment: &v1.EnvironmentConfiguration{
@@ -277,6 +275,7 @@ func TestSelect(t *testing.T) {
 									Selector: &v1.EnvironmentSourceSelector{
 										MatchLabels: []v1.EnvironmentSourceSelectorLabelMatcher{
 											{
+												Type:  v1.EnvironmentSourceSelectorLabelMatcherTypeValue,
 												Key:   "foo",
 												Value: pointer.String("bar"),
 											},
```